orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -3131,3 +3142,199 @@ def test_debug_options_log_level_validation_fails_on_type_mismatch():
+        assert f"Input argument to forward found on device {torch.device(inputs.device)}, but expected it to be on module device {model._torch_module._execution_manager(model._is_training())._device}." in str(e.value)
+
+
+@pytest.mark.parametrize("is_training,fallback_enabled,matching_policy", [(True, True, True), (False, True, True), (True, False, True), (False, False, True),
+                                                                          (True, True, False), (False, True, False), (True, False, False), (False, False, False)])
+def test_ortmodule_fallback_output(is_training, fallback_enabled, matching_policy):
+    # is_training: True for torch.nn.Module training model, eval mode otherwise
+    # fallback_enabled: True results in PyTorch executing the forward graph instead of ORT backend
+    # matching_policy: True results in properly matching FALLBACK_UNSUPPORTED_OUTPUT policy to ORTModuleDeviceException exception.
+    #   Otherwise, an incorrect policy (FALLBACK_UNSUPPORTED_DEVICE) is used to verify that the fallback does not happen
+
+    device = 'cuda'
+    N, D_in, H, D_out = 64, 784, 500, 10
+    pt_model = NeuralNetCustomClassOutput(D_in, H, D_out).to(device)
+    ort_model = ORTModule(pt_model)
+    x = torch.randn(N, D_in, device=device)
+    y = torch.randn(N, D_in, device=device)
+    z = torch.randn(N, D_in, device=device)
+
+    policy = _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_OUTPUT if matching_policy else _fallback._FallbackPolicy.FALLBACK_UNSUPPORTED_DEVICE
+
+    ort_model.train(is_training)
+    pt_model.train(is_training)
+    if fallback_enabled:
+        ort_model._torch_module._execution_manager(is_training=is_training)._fallback_manager._policy = policy
+        if matching_policy:
+            ort_out = ort_model(x, y, z)
+            pt_out = pt_model(x, y, z)
+            _test_helpers.assert_values_are_close(ort_out.out1, pt_out.out1, rtol=0, atol=0)
+            _test_helpers.assert_values_are_close(ort_out.out2, pt_out.out2, rtol=0, atol=0)
+            _test_helpers.assert_values_are_close(ort_out.out3, pt_out.out3, rtol=0, atol=0)
+        else:
+            with pytest.raises(_fallback.ORTModuleIOError) as runtime_error:
+                ort_model(x, y, z)
+            assert 'ORTModule does not support the following model output type' in str(runtime_error.value)
+    else:
+        with pytest.raises(_fallback.ORTModuleIOError) as runtime_error:
+            ort_model(x, y, z)
+        assert 'ORTModule does not support the following model output type' in str(runtime_error.value)
+
+@pytest.mark.parametrize("is_training,fallback_enabled,matching_policy", [(True, True, True), (False, True, True), (True, False, True), (False, False, True),
+                                                                          (True, True, False), (False, True, False), (True, False, False), (False, False, False)])
+def test_ortmodule_fallback_input(is_training, fallback_enabled, matching_policy):
```

Comment:
>def [](http://example.com/codeflow?start=0&length=3)

Let's have a test case where torch-converter fails. 

We can use autograd.function in the a torch.nn.module, and disable ORT pythonOp Support.